### ISSUE TITLE
SD3.5 full-finetune loss nan

### ISSUE THREAD

**Issue by mliand** [kohya-ss/sd-scripts#2104]
![Image](https://github.com/user-attachments/assets/cfb996cc-0ccb-4e42-892a-c0ab7b28ebd4)
On the latest branch of sd3, a full fine-tuning was attempted on sd3.5m. The default parameter was to run loss nan. After inference using deepspeed, the loss would increase rapidly and then nan. It seems that there are serious problems in the training calculation

**Comment by mliand** [comment]
![Image](https://github.com/user-attachments/assets/a307d7d9-87d6-4ac2-ab51-d74a599531cc)
When running on a single gpu, the loss NaN occurs in the second step

**Comment by mliand** [comment]
> What precision you use?

mixed precision bf16

**Comment by deGENERATIVE-SQUAD** [comment]
@mliand 
Are you using optimizers optimized for bf16? Show the full configuration

**Comment by mliand** [comment]
It seems that there is a problem with the gradient backpropagation during the mixed-precision training.

**Comment by mliand** [comment]
I attempted to conduct the training using full BF16. The loss has stabilized. It seems that there was an issue with the gradient backpropagation during the mixed precision training.

**Comment by deGENERATIVE-SQUAD** [comment]
@mliand 

> Moreover, when I attempted to train using fp16, a new error message 'nan found in context. Replacing with zeros' appeared.

The same problem occurs with SDXL as well when using full_fp16. I think it’s some kind of VAE-related bug and should be reported as a separate issue. It can be fixed by using --no_half_vae when training LoRA networks. In the case of SD3, I believe the solution is similar, but since you're training full parameters and the script doesn’t support --no_half_vae to force floats.

Also full_fp16 does not cause errors when using AdamW8bit, give it a try.

> max_grad_norm = 1

At first I think you need to disable clipping when you use AdaFactor, norm is excessive.

AdaFactor can be numerically unstable with your settings, in your case it is like Adagrad now (disables most of the mechanisms that make AdaFactor stable, True modes are default) and bad for large transformers.
If gradient scales aren't managed properly, it can cause overflow/underflow/both of which lead to NaN. Then I would try "scale_parameter=True", "relative_step=True" without warmup and adafactor scheduler to check overall stability.

Additionally, I would check if the problem isn't caused by the optimizer itself. Try using optimizers that are optimized for bf16 instead of AdaFactor: https://github.com/lodestone-rock/torchastic and https://github.com/warner-benjamin/optimi

Have you used anything else unusual apart from max_grad_norm? You said you use deepspeed, can you provide all info about configs, libs, CUDA ver, pytorch ver, etc to make statistic more accurate?

**Comment by mliand** [comment]
I have tried training with adamw. The learning rate is 1e-6, which is very low, but it seems to always be in the nan state，I just simply use data parallelism and run on a single card

**Comment by rockerBOO** [comment]
unfortuntately the gradients are not being reported so makes it hard to know if it's just an exploding gradients problem or something related to that. Those issues are not exactly only training tool related but a combination of the optimizer, dataset, model and other hyperparameters. I do not think there is something built in for full fine tuning that will report the gradients though. 

Example I have for exploding gradients which makes further training cause the loss curve to change dramatically. 
![Image](https://github.com/user-attachments/assets/5448870f-effb-497b-bf03-200e94d2dbf6)

Unfortunately identifying this would be trial and error with the current reporting tooling though, to try to identify what could be causing it to happen (when the loss curve changes more upward).

**Comment by mliand** [comment]
> unfortuntately the gradients are not being reported so makes it hard to know if it's just an exploding gradients problem or something related to that. Those issues are not exactly only training tool related but a combination of the optimizer, dataset, model and other hyperparameters. I do not think there is something built in for full fine tuning that will report the gradients though.
> 
> Example I have for exploding gradients which makes further training cause the loss curve to change dramatically. ![Image](https://github.com/user-attachments/assets/5448870f-effb-497b-bf03-200e94d2dbf6)
> 
> Unfortunately identifying this would be trial and error with the current reporting tooling though, to try to identify what could be causing it to happen (when the loss curve changes more upward).
